### DISCUSSION TITLE
Add support for standard keyboard navigation

### DISCUSSION THREAD

**Discussion by 82MHz** [d99kris/nchat#184] [Ideas]
It would be great if the standard keyboard navigation from most text editors as well as the console would work in the text input window, meaning using the home + end keys to move to the start/end of a line and using ctrl+arrow keys to quickly skip forward/backward through the text one word at a time. 
For reference, I'm on nchat v3.93 on Arch Linux.

**Comment by fulalas** [comment]
It's a matter of setting it correctly. The defaults are not intuitive, I agree.

Edit your key.conf and change these bindings (make sure these key combinations are not set anywhere else):

```
begin_line=KEY_HOME
end_line=KEY_END

backward_word=\33\133\61\73\65\104
forward_word=\33\133\61\73\65\103
```

**Comment by Disonantemus** [comment]
> begin_line=KEY_HOME
> end_line=KEY_END

Those didn't work 4 me.

~~~yaml
$ nchat -v | head -1
nchat v4.13

$ setxkbmap -query
rules:      evdev
model:      pc105
layout:     es
options:    terminate:ctrl_alt_bksp

$ neofetch distro kernel
distro: Arch Linux x86_64
kernel: 6.1.52-1-lts
~~~

**Comment by fulalas** [comment]
In this case you need to post here your whole `key.conf`.

**Reply by d99kris** [reply]
Yes, please share the whole `key.conf` file @Disonantemus - as other key bindings can interfere with what you're trying to set.

**Comment by Disonantemus** [comment]
This is my `key.conf` right now:
~~~ocaml
backspace=KEY_BACKSPACE
backspace_alt=KEY_ALT_BACKSPACE
backward_kill_word=\033\177
backward_word=\33\133\61\73\65\104
begin_line=\4001046
cancel=KEY_CTRLC
clear=KEY_CTRLC
copy=\33\143
cut=\33\170
decrease_list_width=\33\54
delete=KEY_DC
delete_chat=\33\144
delete_line_after_cursor=KEY_CTRLK
delete_line_before_cursor=KEY_CTRLU
delete_msg=KEY_CTRLD
down=KEY_DOWN
edit_msg=KEY_CTRLZ
end=\4001024
end_line=\4001065
ext_call=\33\164
ext_edit=\33\145
forward_word=\33\133\61\73\65\103
home=\4001075
increase_list_width=\33\56
kill_word=\1006
left=KEY_LEFT
next_chat=KEY_TAB
next_page=KEY_NPAGE
open=KEY_CTRLV
open_link=KEY_CTRLW
open_msg=\33\167
other_commands_help=KEY_CTRLO
paste=\33\166
prev_chat=KEY_BTAB
prev_page=KEY_PPAGE
quit=KEY_CTRLQ
return=KEY_RETURN
right=KEY_RIGHT
save=KEY_CTRLR
select_contact=KEY_CTRLN
select_emoji=KEY_CTRLS
send_msg=KEY_CTRLX
spell=\33\163
terminal_focus_in=KEY_FOCUS_IN
terminal_focus_out=KEY_FOCUS_OUT
terminal_resize=KEY_RESIZE
toggle_emoji=KEY_CTRLY
toggle_help=KEY_CTRLG
toggle_list=KEY_CTRLL
toggle_top=KEY_CTRLP
transfer=KEY_CTRLT
unread_chat=KEY_CTRLF
up=KEY_UP
~~~

I settled for now, with this changes to default `key.conf`, because `Home` and `End` keys didn't work (edited to show actual "key names"):
~~~ocaml
$ grep -E 'd_word|[dn]_line|home|^end' ~/.nchat/key.conf
backward_word=\33\133\61\73\65\104    # ^left
begin_line=\4001046                   # A-left
end=\4001024                          # ^down
end_line=\4001065                     # A-right
forward_word=\33\133\61\73\65\103     # ^right
home=\4001075                         # ^up
~~~

Right now, `Home` and `End` shouldn't be assigned, when pressed in `nchat`, shows:

- End: `[4~`
- Home (wait 4 next press to show): `[1~` 

**Keydump** from my keyboard:

~~~yaml
\4171414 home
\4171410 end
\4171403 ^left
\4171413 ^right
\4001075 ^up
\4001024 ^down
\4001046 A-left
\4001065 A-right
\4001062 ^RePag
\4001055 ^AvPag
~~~
I did try to use `^left` and `^right` from this (keydump), but didn't work.
Maybe some functions are **hardcoded** and clash with `key.conf`?

---

- I think that default `home=` and `end=` shouldn't be `Home` and `End` keys, because I never did use it, and I don't see that many people use it that. It's a LOT more useful to assign that to "standard keyboard navigation" as op said.
- Should be a way to assign a function, to at least two keys (because international keyboards has some different keys than `us`).
- Also I did check with [xterm](https://github.com/d99kris/nchat/issues/179#issuecomment-1913625106), and happens the same, with **Meta Send Escape** enabled.
- I don't know if it's related, maybe is another bug?, but when I do `Alt-e` to **edit a message** in [external editor](https://github.com/martanne/vis), also `Home` and `End` don't work, but If I edit outside `nchat` works as expected.

---

Keyboard: [Trust GXT 833 (Spanish)](https://codegeek.es/2021/08/06/gxt-833-thado-tkl-illuminated-gaming-keyboard-review-espanol/)

~~~yaml
$ nchat -v | head -1
nchat v4.13

$ setxkbmap -query
rules:      evdev
model:      pc105
layout:     es
options:    terminate:ctrl_alt_bksp

$ neofetch distro kernel
distro: Arch Linux x86_64
kernel: 6.1.52-1-lts
~~~
